src/main/scala/is/hail/methods/SplitMulti.scala

```diff
@@ -1,158 +1,213 @@
+
+  val gEC = EvalContext(Map(
+    "global" -> (0, vsm.globalSignature),
+    "v" -> (1, vsm.vSignature),
+    "va" -> (2, vsm.vaSignature),
+    "aIndex" -> (3, TInt32),
+    "wasSplit" -> (4, TBoolean),
+    "g" -> (5, vsm.genotypeSignature)))
+  val gAnnotator = new ExprAnnotator(gEC, vsm.genotypeSignature, genotypeExpr, Some(Annotation.GENOTYPE_HEAD))
+
+  val newMatrixType = vsm.matrixType.copy(vaType = vAnnotator.newT, genotypeType = gAnnotator.newT)
+
+  def split(sortAlleles: Boolean, removeLeftAligned: Boolean, removeMoving: Boolean, verifyLeftAligned: Boolean): RDD[RegionValue] = {
+    val localKeepStar = keepStar
+    val localGlobalAnnotation = vsm.globalAnnotation
+    val localNSamples = vsm.nSamples
+    val localRowType = vsm.rowType
+    val localVAnnotator = vAnnotator
+    val localGAnnotator = gAnnotator
+
+    val newRowType = newMatrixType.rowType
+
+    vsm.rdd2.mapPartitions { it =>
+      val context = new SplitMultiPartitionContext(
+        localKeepStar,
+        localNSamples, localGlobalAnnotation, localRowType,
+        localVAnnotator, localGAnnotator, newRowType)
+
+      it.flatMap { rv =>
+        val splitit = context.splitRow(rv, sortAlleles, removeLeftAligned, removeMoving, verifyLeftAligned)
+        context.prevLocus = context.ur.getAs[Locus](0)
+        splitit
       }
-    }.orderedRepartitionBy(vds.rdd.orderedPartitioner)
-
-    val localMaxShift = maxShift
-    val staticVariants = vds.rdd.mapPartitionsWithIndex { case (i, it) =>
-      LocalVariantSortIterator(it.flatMap { case (v, (va, gs)) =>
-        split(v, va, gs,
-          propagateGQ = propagateGQ,
-          keepStar = keepStar,
-          insertSplitAnnots = { (va, index, wasSplit) =>
-            insertSplit(insertIndex(va, index), wasSplit)
-          },
-          f = (v: Variant) => partitionerBc.value.getPartition(v) == i)
-      }, localMaxShift)
     }
+  }
 
-    val newRDD = OrderedRDD.partitionedSortedUnion(staticVariants, shuffledVariants, vds.rdd.orderedPartitioner)
+  def split(): VariantSampleMatrix[Locus, Variant, T] = {
+    val newRDD2: OrderedRDD2 =
+      if (leftAligned) {
+        OrderedRDD2(
+          newMatrixType.orderedRDD2Type,
+          vsm.rdd2.orderedPartitioner,
+          split(sortAlleles = true, removeLeftAligned = false, removeMoving = false, verifyLeftAligned = true))
+      } else {
+        val leftAlignedVariants = OrderedRDD2(
+          newMatrixType.orderedRDD2Type,
+          vsm.rdd2.orderedPartitioner,
+          split(sortAlleles = true, removeLeftAligned = false, removeMoving = true, verifyLeftAligned = false))
+
+        val movedVariants = OrderedRDD2.shuffle(
+          newMatrixType.orderedRDD2Type,
+          vsm.rdd2.orderedPartitioner,
+          split(sortAlleles = false, removeLeftAligned = true, removeMoving = false, verifyLeftAligned = false))
+
+        leftAlignedVariants.partitionSortedUnion(movedVariants)
```

Comment:
so this creates 2 separate rdds of sorted & unsorted elements, orders the second one, and then creates a sorted union?

Reply:
Correct!  The goal is to minimize the size of the shuffled data by only shuffling the things that moved and are therefore (potentially) no longer ordered.  This is in my mind a major improvement over the previous code that tried to re-sort partitions and shuffle only things that moved between partitions.  Re-sorting partitions is really a no-go unless you support off-disk shuffling (as shuffle does) and was a source of memory/failure problems previously that I think this should address.